### PR TITLE
Feature/billink 78 midpage

### PR DESCRIPTION
**Pull request by DarthFly**

### CODE REVIEW

File: Api/MidpagePaymentDataInterface.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+interface MidpagePaymentDataInterface
+{
+    /**
+     * @param int $cartId
+     * @param \Magento\Quote\Api\Data\PaymentInterface $paymentMethod
+     * @param \Magento\Quote\Api\Data\AddressInterface|null $billingAddress
+     * @return \Billink\Billink\Api\MidpageResultDataInterface $object
+     */
+    public function savePaymentInformationAndPlaceOrder(
+        $cartId,
```
Comment (bethlehemit) [inline Api/MidpagePaymentDataInterface.php]: type of cartId

Depending on PHP/M2 version requirements, we can typehint the returntype too

File: Controller/Midpage/AbstractAction.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Framework\Message\ManagerInterface;
+use Magento\Payment\Gateway\Command\CommandPoolInterface;
+use Magento\Payment\Gateway\Data\PaymentDataObjectFactory;
+use Magento\Sales\Api\OrderRepositoryInterface;
+use Psr\Log\LoggerInterface;
+
+abstract class AbstractAction implements HttpGetActionInterface
+{
+    protected RequestInterface $request;
```
Comment (bethlehemit) [inline Controller/Midpage/AbstractAction.php]: if we are on PHP8+, we can use constructor properties (and iirc from 8.1 on, we can even do readonly properties, saves on clutter

File: Controller/Midpage/AbstractAction.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        $this->transactionManager = $transactionManager;
+        $this->commandPool = $commandPool;
+        $this->paymentDataObjectFactory = $paymentDataObjectFactory;
+        $this->messageManager = $messageManager;
+        $this->orderRepository = $orderRepository;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->logger = $logger;
+    }
+
+    protected function getOrderFromTransaction(): ?string
```
Comment (bethlehemit) [inline Controller/Midpage/AbstractAction.php]: getOrderIdFromTransaction

File: Controller/Midpage/AbstractAction.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    protected function loadOrderByIncrementId(string $transactionOrder)
+    {
+        $searchCriteria = $this->searchCriteriaBuilder
+            ->addFilter('increment_id', $transactionOrder)
+            ->create();
+        $orders = $this->orderRepository->getList($searchCriteria)->getItems();
+        $order = array_shift($orders);
+        if (!$order) {
+            throw new LocalizedException(__('Incorrect order number provided: %1. Please reach for the support.', $transactionOrder));
```
Comment (bethlehemit) [inline Controller/Midpage/AbstractAction.php]: *Please reach out to support

File: Controller/Midpage/Cancel.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+class Cancel extends AbstractAction
+{
+    /**
+     * @return
+     */
+    public function execute()
+    {
+        $resultRedirect = $this->redirectFactory->create();
+
+        $transactionOrder = (string)$this->getOrderFromTransaction();
```
Comment (bethlehemit) [inline Controller/Midpage/Cancel.php]: Typecast is unnecessary?

File: Controller/Midpage/Cancel.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        if (!$transactionOrder) {
```
Comment (bethlehemit) [inline Controller/Midpage/Cancel.php]: prefer being explicit -> $transactionOrder === null

File: Controller/Midpage/Cancel.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            $resultRedirect->setPath('checkout/cart');
+            return $resultRedirect;
+        }
+
+        $params = ['_secure' => true];
+        try {
+            $order = $this->checkoutSession->getLastRealOrder();
```
Comment (bethlehemit) [inline Controller/Midpage/Cancel.php:24]: Is this order already loaded, or are we triggering a load. if triggering, might as well skip this and immediatly do ->loadOrderByIncrementId

File: Controller/Midpage/Cancel.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            }
+            $this->paymentSession->deactivatePaymentSessionById($order->getEntityId());
+            /** @var PaymentDataObjectInterface $paymentDO */
+            $paymentDO = [
+                'payment' => $this->paymentDataObjectFactory->create($order->getPayment())
+            ];
+            $command = $this->commandPool->get('order_cancel');
+            $command->execute($paymentDO);
+        } catch (LocalizedException $e) {
+            $this->messageManager->addExceptionMessage($e);
```
Comment (bethlehemit) [inline Controller/Midpage/Cancel.php:37]: Should we log this too?

File: Gateway/Command/MidpageCancelCommand.php [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+     * @param array $commandSubject
+     * @return void
+     * @throws LocalizedException
+     */
+    public function execute(array $commandSubject): void
+    {
+        $paymentDO = SubjectReader::readPayment($commandSubject);
+        /** @var Payment $payment */
+        $payment = $paymentDO->getPayment();
+        ContextHelper::assertOrderPayment($payment);
```
Comment (bethlehemit) [inline Gateway/Command/MidpageCancelCommand.php:34]: can this assert throw? if so, should we catch -> log -> throw again?

File: Gateway/Command/MidpageCaptureCommand.php [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    public function execute(array $commandSubject)
+    {
+        $paymentDO = SubjectReader::readPayment($commandSubject);
+        /** @var Payment $payment */
+        $payment = $paymentDO->getPayment();
+        ContextHelper::assertOrderPayment($payment);
+        try {
+            $order = $payment->getOrder();
+            $invoiceCollection = $order->getInvoiceCollection();
+            $invoice = $invoiceCollection->getFirstItem();
```
Comment (bethlehemit) [inline Gateway/Command/MidpageCaptureCommand.php]: getFirstItem defaults to an empty object on empty collection, so if you want to check if the invoice exists, either check invoiceCollection->getSize() (or count(), but IIRC getSize does not load the object but just runs a COUNT() query)
or check $invoice->getId()

File: Gateway/Command/MidpageGatewayCommand.php [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+     * @param ResultInterface $result
+     */
+    private function processErrors(array $commandSubject, ResultInterface $result)
+    {
+        $payment = SubjectReader::readPayment($commandSubject);
+        $messages = [];
+        $originMessages = [];
+        foreach ($result->getFailsDescription() as $failPhrase) {
+            $message = (string)$failPhrase;
+            $originMessages[] = $message;
```
Comment (bethlehemit) [inline Gateway/Command/MidpageGatewayCommand.php]: naming, is this just messages?

File: Gateway/Request/Midpage/SessionCreate/Transaction.php [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+                'description' => (string)$item->getDescription(),
+                //'productIdentifiers' => [],
+                'totalProductAmount' => (string)$item->getRowTotalInclTax(),
+                'productAmount' => (string)$item->getBasePriceInclTax(),
+                'productTaxAmount' => (string)$item->getTaxAmount(),
+                'taxRate' => (string)$item->getTaxPercent(),
+                'quantity' => (string)$item->getQtyOrdered(),
+            ];
+/*
+            "productIdentifiers": {
```
Comment (bethlehemit) [inline Gateway/Request/Midpage/SessionCreate/Transaction.php]: Cleanup on aisle five

File: view/frontend/web/template/payment/midpage-form.html [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+               class="radio"
+               data-bind="attr: {'id': getCode()}, value: getCode(), checked: isChecked, click: selectPaymentMethod, visible: isRadioButtonVisible()"/>
+        <label data-bind="attr: {'for': getCode()}" class="label"><span data-bind="text: getTitle()"></span></label>
+    </div>
+    <div class="payment-method-content">
+        <!-- ko foreach: getRegion('messages') -->
+        <!-- ko template: getTemplate() --><!-- /ko -->
+        <!--/ko-->
+        <div class="payment-method-billing-address">
+            <!-- ko foreach: $parent.getRegion(getBillingAddressFormName()) -->
```
Comment (bethlehemit) [inline view/frontend/web/template/payment/midpage-form.html:14]: Do we need these KO blocks for the redirect?

**Review by bethlehemit — APPROVED** [review]
Might also be good to run this through PHPmd/PHPcs for formatting and suggestions (Magento also has some custom standards I believe)

Add composer.json requirements for magento/php version?

File: Controller/Midpage/Cancel.php (hunk 5 above)
Comment (DarthFly) [reply Controller/Midpage/Cancel.php]: No, it was before this function was implemented :)

File: Controller/Midpage/AbstractAction.php (hunk 2 above)
Comment (DarthFly) [reply Controller/Midpage/AbstractAction.php]: @bakkerpj  This one is a head-scratcher for me here.
Mostly I thought that this should be PHP7 compatible, but now I see that some functions have `: static` return, which is PHP8. I don't see other places with PHP8-specific, so it almost looks like the extension is still compatible with 7, but not quite.
And the compatibility is the reason I have not used those. So either we drop PHP7 completely (and probably update composer) or we should fix `: static` return.

File: Controller/Midpage/Cancel.php (hunk 7 above)
Comment (DarthFly) [reply Controller/Midpage/Cancel.php:24]: No, this is intended.
Transaction Id above is from the request, while this one validate is we are on the same session as this transaction or IF magento session died and we should finish the order as it is.

File: Gateway/Command/MidpageCancelCommand.php (hunk 9 above)
Comment (DarthFly) [reply Gateway/Command/MidpageCancelCommand.php:34]: Yes, but not sure if we need to log it. If something is broken at this point - it's more for the magento/development flow, than to "something with the data". So this assert makes sure that correct object is provided and that the main class that builds it is not called from some other place with some other params.

File: view/frontend/web/template/payment/midpage-form.html (hunk 13 above)
Comment (DarthFly) [reply view/frontend/web/template/payment/midpage-form.html:14]: Yes, but not all. One is duplicate.

**Comment by DarthFly** [comment]
@bakkerpj  @bethlehemit PR updated, most notices fixed.
The questions about PHP7 support is open.

File: Controller/Midpage/AbstractAction.php (hunk 2 above)
Comment (bakkerpj) [reply Controller/Midpage/AbstractAction.php]: @DarthFly We have explicitly stated: `Versions 2.0.0 and later of this module require Magento 2.4+ and PHP 8.1+.`. So feel free to drop the PHP7 compatibility.

Comment (bakkerpj) [reply Controller/Midpage/AbstractAction.php]: @DarthFly And yes, it is a good idea to enforce this through requirements in the `composer.json`.

File: Controller/Midpage/AbstractAction.php [hunk 14]
```diff
@@ -55,29 +56,25 @@ public function __construct(
...
+    protected function loadOrderByIncrementId(string $transactionOrder): OrderInterface
     {
         $searchCriteria = $this->searchCriteriaBuilder
             ->addFilter('increment_id', $transactionOrder)
             ->create();
         $orders = $this->orderRepository->getList($searchCriteria)->getItems();
         $order = array_shift($orders);
         if (!$order) {
-            throw new LocalizedException(__('Incorrect order number provided: %1. Please reach for the support.', $transactionOrder));
+            throw new LocalizedException(__('Incorrect order number provided: %1. Please reach to support.', $transactionOrder));
```
Comment (bakkerpj) [inline Controller/Midpage/AbstractAction.php]: @DarthFly "...reach out to..."

**Comment by DarthFly** [comment]
@bethlehemit  @bakkerpj  Another update with PHP 8 update.

**Review by bakkerpj — APPROVED** [review]

**Pull request merged**